src/Ormolu/Printer/Meat/Declaration/Value.hs

```diff
@@ -888,33 +889,95 @@ p_listComp s es = sitcc (vlayout singleLine multiLine)
-flattenStmts = collect . map gatherStmt
+-- For example, this code:
+--
+-- @
+-- [ a + b + c + d
+-- | a <- as, let b = a + a
+-- | c <- cs
+-- | d <- ds, then sort by f
+-- ]
+-- @
+--
+-- is parsed as roughly:
+--
+-- @
+-- [ ParStmt
+--     [ ParStmtBlock
+--         [ BindStmt [| a <- as |]
+--         , LetStmt  [| let b = a + a |]
+--         ]
+--     , ParStmtBlock
+--         [ BindStmt [| c <- cs |]
+--         ]
+--     , ParStmtBlock
+--         [ TransStmt
+--             [ BindStmt [| d <- ds |]
+--             ]
+--             [| then sort by f |]
+--         ]
+--     ]
+-- , LastStmt [| a + b + c + d |]
+-- ]
+-- @
+--
+-- The final expression is parsed out in p_body, and the rest is passed
+-- to this function. This function takes the above tree as input and
+-- normalizes it into:
+--
+-- @
+-- [ [ BindStmt [| a <- as |]
+--   , LetStmt  [| let b = a + a |]
+--   ]
+-- , [ BindStmt [| c <- cs |]
+--   ]
+-- , [ BindStmt [| d <- ds |]
+--   , TransStmt [] [| then sortWith by f |]
+--   ]
+-- ]
+-- @
+--
+-- Notes:
+--   * The number of elements in the outer list is the number of pipes in
+--     the comprehension; i.e. 1 unless -XParallelListComp is enabled
+gatherStmts :: [ExprLStmt GhcPs] -> [[ExprLStmt GhcPs]]
+gatherStmts = \case
+  -- When -XParallelListComp is enabled + list comprehension has
+  -- multiple pipes, input will have exactly 1 element, and it
+  -- will be ParStmt.
+  [L _ (ParStmt _ blocks _ _)] ->
+    [ concatMap collectParStmts stmts
+    | ParStmtBlock _ stmts _ _ <- blocks
+    ]
+  -- When -XTransformListComp is enabled and there's only a
+  -- single pipe, the input will have exactly 1 element, and it
+  -- will be TransStmt.
+  [stmt@(L _ TransStmt {trS_stmts})] ->
+    [ map checkTransStmt trS_stmts ++ [stmt]
+    ]
+  -- Otherwise, list will not contain any ParStmt or TransStmt
+  stmts ->
+    [ map checkPlainStmt stmts
+    ]
   where
-    -- Note: not exactly sure what this does...
-    --
-    -- From experimenting, it does this:
-    --
-    -- >>> collect [[[stmt1]], [[stmt2, stmt3]], [[stmt4], [stmt5]], [[stmt6]]]
-    -- [[stmt1, stmt2, stmt3, stmt4, stmt6], [stmt5]]
-    --
-    -- But it's not clear what the purpose is.
-    collect :: [[[ExprLStmt GhcPs]]] -> [[ExprLStmt GhcPs]]
-    collect = foldr (zipPrefixWith (<>)) []
+    collectParStmts = \case
+      L _ ParStmt {} -> unexpected "ParStmt nested inside ParStmt"
+      stmt@(L _ TransStmt {trS_stmts}) -> map checkTransStmt trS_stmts ++ [stmt]
+      stmt -> [stmt]
 
-    gatherStmt :: ExprLStmt GhcPs -> [[ExprLStmt GhcPs]]
-    gatherStmt (L _ (ParStmt _ block _ _)) =
-      concatMap gatherStmtBlock block
-    gatherStmt (L s stmt@TransStmt {..}) =
-      collect $ map gatherStmt trS_stmts <> [[[L s stmt]]]
-    gatherStmt stmt = [[stmt]]
+    checkTransStmt = \case
+      L _ ParStmt {} -> unexpected "ParStmt nested inside TransStmt"
+      L _ TransStmt {} -> unexpected "TransStmt nested inside TransStmt"
```

Comment:
In contrast to `ParStmt`, `TransStmt` can actually occur inside of a `TransStmt`:
```haskell
foo xs = [x | x <- xs, then reverse, then reverse]
```
This will be formatted correctly on `master`/0.7.5.0, but throws an error since your last commit.

One way to fix this would be to apply this diff:
```diff
diff --git a/src/Ormolu/Printer/Meat/Declaration/Value.hs b/src/Ormolu/Printer/Meat/Declaration/Value.hs
index 268cc30..561d6d2 100644
--- a/src/Ormolu/Printer/Meat/Declaration/Value.hs
+++ b/src/Ormolu/Printer/Meat/Declaration/Value.hs
@@ -948,35 +948,18 @@ gatherStmts = \case
   -- multiple pipes, input will have exactly 1 element, and it
   -- will be ParStmt.
   [L _ (ParStmt _ blocks _ _)] ->
-    [ concatMap collectParStmts stmts
-    | ParStmtBlock _ stmts _ _ <- blocks
-    ]
-  -- When -XTransformListComp is enabled and there's only a
-  -- single pipe, the input will have exactly 1 element, and it
-  -- will be TransStmt.
-  [stmt@(L _ TransStmt {trS_stmts})] ->
-    [ map checkTransStmt trS_stmts ++ [stmt]
-    ]
-  -- Otherwise, list will not contain any ParStmt or TransStmt
-  stmts ->
-    [ map checkPlainStmt stmts
+    [ concatMap collectNonParStmts stmts
+      | ParStmtBlock _ stmts _ _ <- blocks
     ]
+  -- Otherwise, list will not contain any ParStmt
+  stmts -> [concatMap collectNonParStmts stmts]
   where
-    collectParStmts = \case
-      L _ ParStmt {} -> unexpected "ParStmt nested inside ParStmt"
-      stmt@(L _ TransStmt {trS_stmts}) -> map checkTransStmt trS_stmts ++ [stmt]
+    collectNonParStmts = \case
+      L _ ParStmt {} -> unexpected "ParStmt"
+      stmt@(L _ TransStmt {trS_stmts}) ->
+        concatMap collectNonParStmts trS_stmts ++ [stmt]
       stmt -> [stmt]
 
-    checkTransStmt = \case
-      L _ ParStmt {} -> unexpected "ParStmt nested inside TransStmt"
-      L _ TransStmt {} -> unexpected "TransStmt nested inside TransStmt"
-      stmt -> stmt
-
-    checkPlainStmt = \case
-      L _ ParStmt {} -> unexpected "ParStmt in list comprehension"
-      L _ TransStmt {} -> unexpected "TransStmt in list comprehension"
-      stmt -> stmt
-
     unexpected label = error $ "Unexpected " <> label <> "! Please file a bug."
 
 p_patSynBind :: PatSynBind GhcPs GhcPs -> R ()
```